neqo-transport/src/connection/mod.rs

```diff
@@ -921,6 +921,13 @@ impl Connection {
         self.streams.cleanup_closed_streams();
     }
 
+    /// Process new input datagrams on the connection.
+    pub fn process_multiple_input(&mut self, dgrams: Vec<Datagram>, now: Instant) {
```

Comment:
I'm not sure what @larseggert specifically meant when he said "array" here, but Rust convention would have this interface as:

```suggestion
    pub fn process_multiple_input(&mut self, dgrams: &[Datagram], now: Instant) {
```

I would also inline the loop in multiple_input.  It's a trivial function that isn't used elsewhere.

Finally, please consider what you will do when `dgrams.is_empty()`.  This function could do nothing in that case.

Reply:
If I change it to `&[Datagram]`, how can I avoid cloning each datagram?

Reply:
Good point.  It's a pretty big change, but I think that you could change all the input/process functions to take a datagram reference instead.  Maybe `impl AsRef<Datagram>` would make the whole thing less of a challenge.

For now, this is fine.

Reply:
I'd like to avoid big change for now. Please let me know if you think it's better to change all input/process functions to take a reference. If so, I'll create another issue.
Thanks.

Reply:
if we want to support GSO/GRO and avoid memory copies, we need to pass a consecutive region of memory in and out of the packet I/O functions: https://blog.cloudflare.com/accelerating-udp-packet-transmission-for-quic

Reply:
Yeah, that is possible here.  The GSO calls happen outside this code and the memory can be cut up and passed in to the stack. That is easier if we move to accepting slices of bytes, but that is more disruptive and we can probably manage with Vec::split_off at first.

Reply:
I'm not sure I follow. On GSO (i.e., TX), the region passed down into the kernel needs to be consecutive. We can obviously copy memory to make it so, but it would be more efficient if that was not needed (or done as a side effect of where we need to copy memory anyway, i.e., when doing TLS encrypt). 

On GRO (i.e., RX) we get a consecutive memory region from the kernel, which we could exploit to get some TLS decrypt efficiency but don't need to.

Reply:
This change was about reading.  If we get a contiguous range of memory, Rust might view that as a single `x: &[u8]`.  It is trivial then for us to split that up into multiple slices `x[..1280], x[1280..2560], ...`, each of which are wrapped in a fresh `Datagram` instance.  We probably couldn't use `Vec` for that, but we could if it was passed as owned memory, like `Vec<u8>`.

I agree that sending is going to be harder.  My original design had the memory provided by the caller, but we decided to keep things this way very early on.  That is still something we can probably manage with a few tweaks to our API, probably just by having the encryption in place, but we could also do in place packet building and save an allocation with only a tiny bit more effort.  We can keep the single-shot nature of the API in that case though, with the caller passing in mutable slices as needed.